### DISCUSSION TITLE
Authenticating with Azure Active Directory

### DISCUSSION THREAD

**Discussion by johntallon** [lowdefy/lowdefy#1139] [Snippets]
Hi there, 

I am hoping you can help. Looking to authenticate the app against Azure Active Directory, which I am familiar with when working with SPA built with Angular for example. 

Can you provide guidance for hooking up to Azure AD? There are a number of sample for [JavaScript apps available for Azure](https://docs.microsoft.com/en-us/azure/active-directory/develop/single-page-app-quickstart?pivots=devlang-javascript), but I am not entirely sure how that will work in the context of Lowdefy and how to config the solution. 

Any help, would be wonderful... under pressure to show case how Lowdefy could help out here!

**Comment by SamTolmay** [comment]
Hi @johntallon 

I had a look at the Azure Active Directory documentation, and didn't find a clear guide on how to set up an OpenID Connect application. This is the best resource I can find: https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-web-app-sign-user-app-registration?tabs=nodejs

You need to look at the guides for web apps, and not SPA's, since Lowdefy has a backend server to handle the authentication, where the SPA guides use an authentication flow that only runs on the frontend, and does not use the server.

To setup the OpenID  client, you need to configure 4 variables:
```
LOWDEFY_SECRET_OPENID_CLIENT_ID = YOUR_CLIENT_ID
LOWDEFY_SECRET_OPENID_CLIENT_SECRET = YOUR_CLIENT_SECRET
LOWDEFY_SECRET_OPENID_DOMAIN = YOUR_OPENID_DOMAIN
LOWDEFY_SECRET_JWT_SECRET = RANDOM_SECRET
```

The linked guide (on the Node.js tab) gives details on how to obtain the  OPENID_CLIENT_ID and OPENID_CLIENT_SECRET. The JWT_SECRET should be rendomly generated, we have a guide in our docs. 

It looks like you might be able to use `https://login.microsoftonline.com/common` as the  OPENID_DOMAIN.

You could maybe also have a look at these docs: https://docs.pivotal.io/p-identity/1-14/azure-oidc/config-azure.html
They use `https://login.microsoftonline.com/TENANT/v2.0/.well-known/openid-configuration` as the domain, where `TENANT` is the "Tenant ID" or "Primary domain".


You will also need to configure the redirect URL in active directory. These are urls that the user will be redirected to after logging in. You need to specifiy these so the user is not directed to an unauthorized page by an attacker. For Lowdefy, the redirect URL is `http://localhost:3000/auth/openid-callback` for development and `https://app.example.com/auth/openid-callback` for production apps.

Hope this helps

**Reply by johntallon** [reply]
@SamTolmay 

Adding the LOWDEFY_SECRET_OPENID_DOMAIN as https://login.microsoftonline.com/common and adding the Web App platform to the App Reg in Azure fixed the issue. 

Thanks for your response on this, that was very helpful.
